src/connection.rs

```diff
@@ -1983,259 +1915,349 @@ where
 #[allow(clippy::too_many_arguments)]
-async fn server_req_handler<S: AsyncRead + AsyncWrite>(
+async fn server_req_read_body<R: AsyncRead, W: AsyncWrite>(
     id: &str,
-    stream: &mut S,
+    req: &http1::Request<'_, '_>,
+    req_body: &mut server::RequestBodyKeepHeader<'_, '_, R, W>,
     peer_addr: Option<&SocketAddr>,
     secure: bool,
-    buf1: &mut VecRingBuffer,
-    buf2: &mut VecRingBuffer,
     body_buf: &mut ContiguousBuffer,
     packet_buf: &RefCell<Vec<u8>>,
-    zsender: &AsyncLocalSender<zmq::Message>,
     zreceiver: &TrackedAsyncLocalReceiver<'_, (arena::Rc<zhttppacket::OwnedResponse>, usize)>,
-) -> Result<bool, Error> {
-    let stream = RefCell::new(stream);
-
-    let handler = RequestHandler::new(io_split(&stream), buf1, buf2);
-    let mut scratch = http1::ParseScratch::<HEADERS_MAX>::new();
-    let mut req_mem = None;
-
-    // receive request header
-
-    // ABR: discard_while
-    let handler = match discard_while(
-        zreceiver,
-        pin!(handler.recv_request(&mut scratch, &mut req_mem)),
-    )
-    .await
-    {
-        Ok(handler) => handler,
-        Err(Error::Io(e)) if e.kind() == io::ErrorKind::UnexpectedEof => return Ok(false),
-        Err(e) => return Err(e),
-    };
-
-    // log request
+) -> Result<zmq::Message, Error> {
+    // receive request body
 
-    {
-        let req = handler.request();
-        let host = get_host(req.headers);
-        let scheme = if secure { "https" } else { "http" };
+    loop {
+        match req_body.try_recv(body_buf.write_buf())? {
+            RecvStatus::Complete((), written) => {
+                body_buf.write_commit(written);
+                break;
+            }
+            RecvStatus::Read((), written) => {
+                body_buf.write_commit(written);
 
-        debug!(
-            "server-conn {}: request: {} {}://{}{}",
-            id, req.method, scheme, host, req.uri
-        );
+                if written == 0 {
+                    // ABR: discard_while
+                    discard_while(zreceiver, pin!(req_body.add_to_buffer())).await?;
+                }
+            }
+        }
     }
 
-    // receive request body
-
-    // ABR: discard_while
-    let handler = discard_while(zreceiver, pin!(handler.start_recv_body_and_keep_header())).await?;
+    // determine how to respond
 
-    loop {
-        // ABR: discard_while
-        let size = discard_while(zreceiver, pin!(handler.recv_body(body_buf.write_buf()))).await?;
+    let mut websocket = false;
 
-        if size == 0 {
+    for h in req.headers.iter() {
+        if h.name.eq_ignore_ascii_case("Upgrade") && h.value == b"websocket" {
+            websocket = true;
             break;
         }
-
-        body_buf.write_commit(size);
     }
 
-    // determine how to respond
+    if websocket {
+        // websocket requests are not supported in req mode
 
-    let msg = {
-        let req = handler.request();
+        // toss the request body
+        body_buf.clear();
 
```

Comment:
do we wanna add a log line here?

Reply:
previously, the req mode handler would return an error response to a websocket request, which should be sufficient for visibility. I've temporarily removed that handling here in order to consolidate it with other error responses, and so it'll come back in another PR shortly after.